File: Regression/Multiple_linear_regression/multiple_linear_regression.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jul 17 16:19:44 2019

@author: alexa
"""

# Multiple Linear Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Creating dummy variables
dummies = pd.get_dummies(X[:, -1])
X = np.concatenate((dummies, X[:, 0:3]), axis = 1).astype(float)

# Avoiding the dummy variable trap
X = X[:, 1:]

# Splitting the dataset into training set and test set
from sklearn.model_selection import train_test_split
# test_size is a small percentage = 20% - 40%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2,
                                                    random_state = 0)
# Feature Scaling
#from sklearn.preprocessing import StandardScaler
#sc_X = StandardScaler()
#X_train = sc_X.fit_transform(X_train)
#X_test = sc_X.transform(X_test)

# Fitting Multiple Linear Regression to the Training Set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predicting the Test set results
y_pred = regressor.predict(X_test)

# Building the optimal model using Backward Elimination
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50, 1), dtype = int), values = X, axis = 1)

# All-in
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# p of x2 = 0.99

# remove x2
X_opt = X[:, [0, 1, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# p of x1 = 0.94

# remove x1
X_opt = X[:, [0, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# p of col4 (admin) = 0.602

# remove admin col4
X_opt = X[:, [0, 3, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# p of col5 (marketing spend) is = 0.06
# Adj R2 = 0.948

# remove marketing col5
X_opt = X[:, [0, 3]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# all p-values are 0
# Adj R2 = 0.945
# thus, keep the prev model

# Final model
X_opt = X[:, [0, 3, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()

File: Regression/Multiple_linear_regression/multiple_linear_regression.R
# Multiple Linear Regression

# importing dataset
dataset = read.csv("50_Startups.csv")
# dataset = dataset[, 2:3]

# Encoding categorical data
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1, 2, 3))

# Splitting the dataset into training set and test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)

training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

# Fitting the multiple linear regression to the training set
regressor = lm(formula = Profit ~ .,
               data = training_set)

# Predicting the training set results
y_pred = predict(regressor, newdata = test_set)

# Backward elimination
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               data = dataset)
summary(regressor)

# remove state
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)

# remove admin
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)
# adj R2 = 0.948 best

# remove marketing spend
regressor = lm(formula = Profit ~ R.D.Spend,
               data = dataset)
summary(regressor)

# final model
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor)